ssl/quic/quic_sf_list.c

```diff
@@ -0,0 +1,250 @@
+    UINT_RANGE range;
+    OSSL_QRX_PKT_WRAP *pkt;
+    const unsigned char *data;
+};
+
+static void stream_frame_free(SFRAME_LIST *fl, STREAM_FRAME *sf)
+{
+    ossl_qrx_pkt_wrap_free(fl->qrx, sf->pkt);
+    OPENSSL_free(sf);
+}
+
+static STREAM_FRAME *stream_frame_new(UINT_RANGE *range, OSSL_QRX_PKT_WRAP *pkt,
+                                      const unsigned char *data)
+{
+    STREAM_FRAME *sf = OPENSSL_zalloc(sizeof(*sf));
+
+    if (pkt != NULL && !ossl_qrx_pkt_wrap_up_ref(pkt)) {
+        OPENSSL_free(sf);
+        return NULL;
+    }
+
+    sf->range = *range;
+    sf->pkt = pkt;
+    sf->data = data;
+
+    return sf;
+}
+
+void ossl_sframe_list_init(SFRAME_LIST *fl, OSSL_QRX *qrx)
+{
+    memset(fl, 0, sizeof(*fl));
+    fl->qrx = qrx;
+}
+
+void ossl_sframe_list_destroy(SFRAME_LIST *fl)
+{
+    STREAM_FRAME *sf, *next_frame;
+
+    for (sf = fl->head; sf != NULL; sf = next_frame) {
+        next_frame = sf->next;
+        stream_frame_free(fl, sf);
+    }
+}
+
+static int append_frame(SFRAME_LIST *fl, UINT_RANGE *range,
+                        OSSL_QRX_PKT_WRAP *pkt,
+                        const unsigned char *data)
+{
+    STREAM_FRAME *new_frame;
+
+    if ((new_frame = stream_frame_new(range, pkt, data)) == NULL)
+        return 0;
+    new_frame->prev = fl->tail;
+    if (fl->tail != NULL)
+        fl->tail->next = new_frame;
+    fl->tail = new_frame;
+    ++fl->num_frames;
+    return 1;
+}
+
+int ossl_sframe_list_insert(SFRAME_LIST *fl, UINT_RANGE *range,
+                            OSSL_QRX_PKT_WRAP *pkt,
+                            const unsigned char *data, int fin)
+{
+    STREAM_FRAME *sf, *new_frame, *prev_frame, *next_frame;
+#ifndef NDEBUG
+    uint64_t curr_end = fl->tail != NULL ? fl->tail->range.end
+                                         : fl->offset;
+
+    /* This check is handled by QUIC FC already */
+    if ((fin && curr_end > range->end)
+        || (fl->fin && curr_end < range->end)) {
+        /* protocol violation: FINAL_SIZE_ERROR */
+        return 0;
+    }
+#endif
+
+    /* nothing there yet */
+    if (fl->tail == NULL) {
+        if (fl->offset >= range->end)
+            goto end;
+
+        fl->tail = fl->head = stream_frame_new(range, pkt, data);
+        if (fl->tail == NULL)
+            return 0;
+        ++fl->num_frames;
+        goto end;
+    }
+
+    if (fl->offset >= range->end)
+        goto end;
+
+    /* TODO(QUIC): Check for fl->num_frames and start copying if too many */
+
+    /* optimize insertion at the end */
+    if (fl->tail->range.start < range->start) {
+        if (fl->tail->range.end >= range->end) {
+            goto end;
+        }
+        return append_frame(fl, range, pkt, data);
+    }
+
+    prev_frame = NULL;
+    for (sf = fl->head; sf != NULL && sf->range.start < range->start;
+         sf = sf->next)
+        prev_frame = sf;
+
+    if (prev_frame != NULL && prev_frame->range.end >= range->end) {
+        goto end;
+    }
+
+    if (sf == NULL) {
```

Comment:
Can this ever happen? Isn't this already handled by the "optimize insertion at the end" check above?

Reply:
Yes, right. I'll make it an assert.